AppInfoForm update with new fields and validation

st-jay opened the pull request:
## Summary of change
Updated appInfoForm

## Related issues
- https://github.com/supertokens/main-website/issues/585

## Checklist
- [ ] Algolia search needs to be updated? (If there is a new sub docs project, then yes)
- [ ] Sitemap needs to be updated? (If there is a new sub docs project, then yes)
- [ ] Checked for broken links? (Run `cd v2 && MODE=production npx docusaurus build`)
- [ ] Changes required to the demo apps corresponding to the docs?

## Remaining TODOs for this PR
- [X] Add a link at the top of form saying `To learn more about appInfo, read here.`
- [X] Add fields for `apiBasePath` and `websiteBasePath`
- [X] Added a checkbox for NextJS docs for NextJS API Routes
- [X] Updated validation for `apiDomain` and `websiteDomain`
- [X] Add validations for `apiBasePath` and `websiteBasePath` to see if it follows the path pattern: `/some/path`
- [ ] Update `AppInfoForm` component in all docs

st-jay commented:
## Updated pages

### emailpassword

NestJS
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/emailpassword/nestjs/guide#adding-the-module-to-the-application

NextJS
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/emailpassword/nextjs/init#3-create-the-appinfo-configuration
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/emailpassword/nextjs/init#3-create-the-appinfo-configuration

Quick Setup ( backend )
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/emailpassword/quick-setup/backend#2-initialise-supertokens

Quick Setup ( frontend )
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/emailpassword/quick-setup/frontend#2-call-the-init-function


### session

NestJS
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/session/nestjs/guide#adding-the-module-to-the-application

Quick Setup ( backend )
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/session/quick-setup/backend#2-initialise-supertokens

Quick Setup ( frontend )
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/session/quick-setup/frontend#2-call-the-init-function


### thirdparty

NestJS
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/thirdparty/nestjs/guide#adding-the-module-to-the-application
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/thirdparty/nestjs/guide#3-adding-a-service

NextJS
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/thirdparty/nextjs/init#3-create-the-appinfo-configuration
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/thirdparty/nextjs/init#5-create-a-backend-config-function

Quick Setup ( backend )
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/thirdparty/quick-setup/backend#2-initialise-supertokens
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/thirdparty/quick-setup/backend#3-initialise-social-login-providers

Quick Setup ( frontend )
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/thirdparty/quick-setup/frontend#2-call-the-init-function


### thirdpartryemailpassword

NestJS
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/thirdpartyemailpassword/nestjs/guide#adding-the-module-to-the-application
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/thirdpartyemailpassword/nestjs/guide#3-adding-a-service

NextJS
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/thirdpartyemailpassword/nextjs/init#3-create-the-appinfo-configuration
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/thirdpartyemailpassword/nextjs/init#5-create-a-backend-config-function

Quick Setup ( backend )
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/thirdpartyemailpassword/quick-setup/backend#2-initialise-supertokens
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/thirdpartyemailpassword/quick-setup/backend#3-initialise-social-login-providers

Quick Setup ( frontend )
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/thirdpartyemailpassword/quick-setup/frontend#2-call-the-init-function

st-jay commented:
@nkshah2 please check out the forms at the following locations. They are not updated yet, please let me know if and how I should update them.


### emailpassword
Netlify
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/emailpassword/serverless/with-netlify/frontend#2-add-apibasepath-in-the-appinfo-object
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/emailpassword/serverless/with-netlify/backend-config#3-create-a-backend-config-function

AWS Lambda
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/emailpassword/serverless/with-aws-lambda/backend-config#3-create-a-backend-config-function

### session
Netlify
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/session/serverless/with-netlify/frontend#2-add-apibasepath-in-the-appinfo-object
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/session/serverless/with-netlify/backend-config#3-create-a-backend-config-function

AWS Lambda
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/session/serverless/with-aws-lambda/backend-config#3-create-a-backend-config-function

### thirdparty
Netlify
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/thirdparty/serverless/with-netlify/frontend#2-add-apibasepath-in-the-appinfo-object
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/thirdparty/serverless/with-netlify/backend-config#3-create-a-backend-config-function

AWS Lambda
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/thirdparty/serverless/with-aws-lambda/backend-config#3-create-a-backend-config-function

### thirdpartyemailpassword
Netlify
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/thirdpartyemailpassword/serverless/with-netlify/frontend#2-add-apibasepath-in-the-appinfo-object
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/thirdpartyemailpassword/serverless/with-netlify/backend-config#3-create-a-backend-config-function

AWS Lambda
- https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/thirdpartyemailpassword/serverless/with-aws-lambda/backend-config#3-create-a-backend-config-function

nkshah2 commented:
@st-jay Ive ran your changes locally to verify them, theres a few issues that need to be fixed:

- Everywhere the form is being used `apiDomain` and `websiteDomain` should be mandatory and both base paths should be optional (with default values). Currently its the other way around
- The text at the top of the form should change to "To learn more about what these properties mean read *here*" with *here* being the link
- In the form the title and the text about appInfo are too close together making it look cluttered. Add some pace between the,
- For NextJS docs, the default value for apiDomain with the checkbox checked should be `/api/auth` and not `auth`. Right now the user would always get an error for the default values which does not make sense

There probably are more issues but I havent taken a proper look because I found fundamental ones, so please double check everything once

nkshah2 commented on v2/emailpassword/nestjs/guide.mdx:
```diff
@@ -102,9 +102,12 @@ export class AuthModule implements NestModule {
 ### Adding the module to the application
 
```
Refer to the comment on the PR

st-jay replied:
updated

nkshah2 commented:
On a side note (this is more of an enhancement), the texts that show on hover add no real value to the form because even after reading them the properties dont make any more sense. If we can place the app info link better we may not need to have the hover boxes at all. Ill think on this more but just food for thought

st-jay commented:
In the ticket it was mentioned that the user should be allowed to have empty values for domains @nkshah2

nkshah2 commented:
> In the ticket it was mentioned that the user should be allowed to have empty values for domains @nkshah2

Ah, youre right. It should have said users are allowed to have empty values for both base paths and not domains. Ive updated the issue description

rishabhpoddar commented:
## Feedback (in progress)
- [ ] The copy "This is the base path for your API" and "This is the base path for your website" are bad. Change to something like "SuperTokens will expose it's APIs scoped by this base API path" and "SuperTokens UI will be shown on this website route"
- [ ] These seems to be a bug (the invalid error is showing against the wrong field). Same goes for if the website domain is valid, but the API domain is invalid. I think it doesn't consider `http://localhost` as a valid domain. But that should be:
<img width="826" alt="Screenshot 2021-12-24 at 11 29 34" src="https://user-images.githubusercontent.com/2976287/147322652-53c76bd0-48cc-4212-a6bc-e49b4e0be66c.png">
 
- [ ] If I enter just `a.com` in the website / API domain, it complains about "should have a valid protocol". But earlier, we would just add https:// to such cases. Just like we do in the actual SDK. Please make sure to follow the SDK website / API domain normalisation rule in this widget too!
- [ ] In plain JS and react native code tab, we don't add the apiBasePath.. why not?
- [ ] In python, ` api_base_path: "/auth",` is incorrect. It should be with an `=` to sign. Same goes for `website_base_path`
- [ ]

nkshah2 commented:
@st-jay 

- [ ] When you hover next to API domain it says "This is the URL of your app's API domain" which makes no sense. The copy can be changed to "This is the URL of your apps API server" or something similar
- [ ] For frontend quick setup step 5, it should not say "visit /auth" because that can be misleading now that we allow users to provide a base path. Instead it should use the base path from the form values (not sure how complicated this is)
- [ ] In backend quick setup where we explain "This middleware adds a few APIs", the API routes should use the correct base path (If this is doable then the line '"/auth" is a base path that can be changed...' is no longer needed
- [ ] NextJS section numbering is messed up
- [ ] In NextJS backend config step, is the info form required here? Since we moved appInfo into its own section now and the code snippet does not use any of the values set in the form we can remove the form from here (@rishabhpoddar ).
- [ ] NextJS showing the login UI: Step 3 should use the correct api base path
- [ ] The form in general throws an invalid URL error if I use `.` in any of the paths (for both API and Website base paths). For example `https://test.io/.path/path` is a valid Javascript URL but using `/.path/path` as any of the base paths gives me an error. Additionally, in Netlify `/.netlify/functions/path/path` is a valid entry but `/.netlify/functions/.path/path` is not

@rishabhpoddar @st-jay Since we are linking to the appInfo page from the form directly now, can we consider making the design for the link better and dropping the hover elements? The copies in the hover dont really add any value in my opinion

st-jay commented:
### Changes tested from feedback
- [x] When you hover next to API domain it says "This is the URL of your app's API domain" which makes no sense. The copy can be changed to "This is the URL of your apps API server" or something similar
- [x] For frontend quick setup step 5, it should not say "visit /auth" because that can be misleading now that we allow users to provide a base path. Instead it should use the base path from the form values (not sure how complicated this is)
- [x] In backend quick setup where we explain "This middleware adds a few APIs", the API routes should use the correct base path (If this is doable then the line '"/auth" is a base path that can be changed...' is no longer needed
- [x] NextJS section numbering is messed up
- [x] In NextJS backend config step, is the info form required here? Since we moved appInfo into its own section now and the code snippet does not use any of the values set in the form we can remove the form from here (@rishabhpoddar ).
  - **Response: Not needed to change in this PR**
- [x] NextJS showing the login UI: Step 3 should use the correct api base path
- [x] The form in general throws an invalid URL error if I use `.` in any of the paths (for both API and Website base paths). For example `https://test.io/.path/path` is a valid Javascript URL but using `/.path/path` as any of the base paths gives me an error. Additionally, in Netlify `/.netlify/functions/path/path` is a valid entry but `/.netlify/functions/.path/path` is not
- [x] The copy "This is the base path for your API" and "This is the base path for your website" are bad. Change to something like "SuperTokens will expose it's APIs scoped by this base API path" and "SuperTokens UI will be shown on this website route"
- [x] These seems to be a bug (the invalid error is showing against the wrong field). Same goes for if the website domain is valid, but the API domain is invalid. I think it doesn't consider `http://localhost` as a valid domain. But that should be:
<img width="826" alt="Screenshot 2021-12-24 at 11 29 34" src="https://user-images.githubusercontent.com/2976287/147322652-53c76bd0-48cc-4212-a6bc-e49b4e0be66c.png">
 
- [x] If I enter just `a.com` in the website / API domain, it complains about "should have a valid protocol". But earlier, we would just add https:// to such cases. Just like we do in the actual SDK. Please make sure to follow the SDK website / API domain normalisation rule in this widget too!
- [x] In plain JS and react native code tab, we don't add the apiBasePath.. why not?
- [x] In python, ` api_base_path: "/auth",` is incorrect. It should be with an `=` to sign. Same goes for `website_base_path`
- [x] Do we enforce in app info that if we are showing websiteDomain, we should also show websiteBasePath field? (and sabe for API domain and API base path)? So you never need to say `askForAPIDomain` and `askForAPIBasePath` in props for the component since one always implies the other.
   - Side note: In case of NextJS then, we will end up showing the API domain as well even if that checkbox is checked. In this case, if they edit the website domain, that edit should be reflected in the API domain field too (in real time), and vice versa
   - **Response: Modified the code to use `askForAPIBasePath` and `askForWebsiteBasePath` only when we need the `apiBasePath` or the `websiteBasePath` without its domain counterpart. If the `askForAPIDomain` or `askForWebsiteDomain` is true, the basePaths are shown by default without any extra flag.**
- [x] If I uncheck `I am using NextJS`, then the prefilled apiBasePath is still `/api/auth`. It should be `/auth` if that is unchecked (by default).
- [x] NextJS's app info -> code seems to be very buggy. The resulting code doesn't have `websiteBasePath`, and the `apiDomain` is an empty string... If this is an error due to the login in appInfo form, it shows that something is seriously messed up in there in terms of code flow. Please reevaluate / redo!!
- [x] The contents of `/docs/thirdpartyemailpassword/serverless/with-netlify/frontend` don't make sense anymore. Since it's saying the value of apiBasePath should be `/.netlify/functions/auth` regardless of what I enter as the API base path elsewhere. The actual flow here should be similar to how we do for nextJS where we have a checkbox and if the user checks that, it should prepend `/.netlify/functions/` in the API base path.
- [x] In AWS lambda, in the appInfo, we don't consider API base path and website base path? In general, we must always show website base path if we are showing website domain. Same goes for api base path and api domain.

nkshah2 commented:
- [ ] Leaving the base path (website or api) results in the code snippet and content using `""` as the value. The default value should be `"/auth"`. This also results in other problems such as:
    - Path for showing the Login UI is wrongly displayed as `"/"` when it should be `"/auth"` by default
    - Third party callback URLS do not include any base paths (this would result in the login flows to not work)
- [ ] In frontend quick setup -> Step 3 -> with router dom: The code snippet has a comment explaining that the login UI is rendered on the `"/auth"` route regardless of what value is entered in the form. If we cannot use the form value in the snippet we should change the comment to explain something along the lines of `"The login UI is displayed on the websiteBasePath (/auth by default) that is configured when initialising SuperTokens"`
- [ ] Common Customisations -> Changing Base Path: The content will need to change slightly because we now use the form value for base paths and dont default to `"/auth"` everywhere in the docs. Create an issue for this so we can work on it in the future (If you have an issue for enhancements already just add this as a point there)
- [ ] In NextJS with the checkbox is checked the form is allowed to submit without any value for apiBasePath. In this case the base path is displayed as `""` which is incorrect. Unchecking and checking then displays `"/api/auth"` as the base path but this behaviour should also be true when submitting without a value
- [ ] NextJS -> Configuration -> Step 5: The form adds no value here because the code snippet does not use any of the values in it. 
- [ ] Netlify docs: The form should explain that the api base path must include `"/.netlify/functions"` in the error message when the checkbox is checked. The current error makes it seem that the entered value is in invalid path
- [ ] Netlify docs: the form has the same problem as everywhere else when submitting with an empty value for the base paths
- [ ] Netlify docs do not use the entered values for the callback URLs in third party providers
- [ ] AWS Lambda docs do not use the entered values for the callback URLs in third party providers

@st-jay I checked just the third party email password recipe for these (not the others because I found issues here). Please cross check behaviour for all recipes once

st-jay commented:
@nkshah2 
### Changes tested from feedback
- [x] The copy "This is the base path for your API" and "This is the base path for your website" are bad. Change to something like "SuperTokens will expose it's APIs scoped by this base API path" and "SuperTokens UI will be shown on this website route"
- [x] These seems to be a bug (the invalid error is showing against the wrong field). Same goes for if the website domain is valid, but the API domain is invalid. I think it doesn't consider `http://localhost` as a valid domain. But that should be:
<img width="826" alt="Screenshot 2021-12-24 at 11 29 34" src="https://user-images.githubusercontent.com/2976287/147322652-53c76bd0-48cc-4212-a6bc-e49b4e0be66c.png">
 
- [x] If I enter just `a.com` in the website / API domain, it complains about "should have a valid protocol". But earlier, we would just add https:// to such cases. Just like we do in the actual SDK. Please make sure to follow the SDK website / API domain normalisation rule in this widget too!
- [x] In plain JS and react native code tab, we don't add the apiBasePath.. why not?
- [x] In python, ` api_base_path: "/auth",` is incorrect. It should be with an `=` to sign. Same goes for `website_base_path`
- [x] Do we enforce in app info that if we are showing websiteDomain, we should also show websiteBasePath field? (and sabe for API domain and API base path)? So you never need to say `askForAPIDomain` and `askForAPIBasePath` in props for the component since one always implies the other.
   - Side note: In case of NextJS then, we will end up showing the API domain as well even if that checkbox is checked. In this case, if they edit the website domain, that edit should be reflected in the API domain field too (in real time), and vice versa
   - **Response: Modified the code to use  `askForAPIBasePath`  and  `askForWebsiteBasePath`  only when we need the  `apiBasePath`  or the  `websiteBasePath`  without its domain counterpart. If the  `askForAPIDomain`  or  `askForWebsiteDomain`  is true, the basePaths are shown by default without any extra flag.**
- [x] If I uncheck `I am using NextJS`, then the prefilled apiBasePath is still `/api/auth`. It should be `/auth` if that is unchecked (by default).
- [x] NextJS's app info -> code seems to be very buggy. The resulting code doesn't have `websiteBasePath`, and the `apiDomain` is an empty string... If this is an error due to the login in appInfo form, it shows that something is seriously messed up in there in terms of code flow. Please reevaluate / redo!!
- [x] The contents of `/docs/thirdpartyemailpassword/serverless/with-netlify/frontend` don't make sense anymore. Since it's saying the value of apiBasePath should be `/.netlify/functions/auth` regardless of what I enter as the API base path elsewhere. The actual flow here should be similar to how we do for nextJS where we have a checkbox and if the user checks that, it should prepend `/.netlify/functions/` in the API base path.
- [x] In AWS lambda, in the appInfo, we don't consider API base path and website base path? In general, we must always show website base path if we are showing website domain. Same goes for api base path and api domain.
- [x] When you hover next to API domain it says "This is the URL of your app's API domain" which makes no sense. The copy can be changed to "This is the URL of your apps API server" or something similar
- [x] For frontend quick setup step 5, it should not say "visit /auth" because that can be misleading now that we allow users to provide a base path. Instead it should use the base path from the form values (not sure how complicated this is)
- [x] In backend quick setup where we explain "This middleware adds a few APIs", the API routes should use the correct base path (If this is doable then the line '"/auth" is a base path that can be changed...' is no longer needed
- [x] NextJS section numbering is messed up
- [x] In NextJS backend config step, is the info form required here? Since we moved appInfo into its own section now and the code snippet does not use any of the values set in the form we can remove the form from here (@rishabhpoddar ).
	- The form is not to be removed in this PR. 
- [X] NextJS showing the login UI: Step 3 should use the correct api base path
	- Not to be changed for now.
- [x] The form in general throws an invalid URL error if I use `.` in any of the paths (for both API and Website base paths). For example `https://test.io/.path/path` is a valid Javascript URL but using `/.path/path` as any of the base paths gives me an error. Additionally, in Netlify `/.netlify/functions/path/path` is a valid entry but `/.netlify/functions/.path/path` is not
- [x] Leaving the base path (website or api) results in the code snippet and content using `""` as the value. The default value should be `"/auth"`. This also results in other problems such as:
    - Path for showing the Login UI is wrongly displayed as `"/"` when it should be `"/auth"` by default
    - Third party callback URLS do not include any base paths (this would result in the login flows to not work)
- [x] In frontend quick setup -> Step 3 -> with router dom: The code snippet has a comment explaining that the login UI is rendered on the `"/auth"` route regardless of what value is entered in the form. If we cannot use the form value in the snippet we should change the comment to explain something along the lines of `"The login UI is displayed on the websiteBasePath (/auth by default) that is configured when initialising SuperTokens"`
- [x] Common Customisations -> Changing Base Path: The content will need to change slightly because we now use the form value for base paths and dont default to `"/auth"` everywhere in the docs. Create an issue for this so we can work on it in the future (If you have an issue for enhancements already just add this as a point there)
	- Added to https://github.com/supertokens/docs/issues/206
- [x] In NextJS with the checkbox is checked the form is allowed to submit without any value for apiBasePath. In this case the base path is displayed as `""` which is incorrect. Unchecking and checking then displays `"/api/auth"` as the base path but this behaviour should also be true when submitting without a value
- [x] NextJS -> Configuration -> Step 5: The form adds no value here because the code snippet does not use any of the values in it. 
	- The form is not to be removed in this PR.
- [x] Netlify docs: The form should explain that the api base path must include `"/.netlify/functions"` in the error message when the checkbox is checked. The current error makes it seem that the entered value is in invalid path
- [x] Netlify docs: the form has the same problem as everywhere else when submitting with an empty value for the base paths
- [x] Netlify docs do not use the entered values for the callback URLs in third party providers
- [x] AWS Lambda docs do not use the entered values for the callback URLs in third party providers
- [x] please also merge master into this branch and make sure you have made changes to passwordless recipe as well

rishabhpoddar commented:
# Feedback
- [ ] If there is nothing stored in localstorage, some sections (that require just website base path or just API base path), still show that the form has been submitted. It should show the form instead.
- [ ] Some sections that don't have the form in production, have the form in this PR - like "Setup the login UI" and "View the login UI (ReactJS only)" setup in [here](https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/passwordless/quick-setup/frontend). Why?
- [ ] If I enter a path in API domain or website domain it just gets ignored. I thought we were going to show an error to the user saying that they should instead use API base path / website base path right? Makes me wonder what else is broken...

st-jay commented:
> # Feedback
> * [ ]  If there is nothing stored in localstorage, some sections (that require just website base path or just API base path), still show that the form has been submitted. It should show the form instead.
> * [ ]  Some sections that don't have the form in production, have the form in this PR - like "Setup the login UI" and "View the login UI (ReactJS only)" setup in [here](https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/passwordless/quick-setup/frontend). Why?
> * [ ]  If I enter a path in API domain or website domain it just gets ignored. I thought we were going to show an error to the user saying that they should instead use API base path / website base path right? Makes me wonder what else is broken...

- [X]  If there is nothing stored in localstorage, some sections (that require just website base path or just API base path), still show that the form has been submitted. It should show the form instead.
- [X]  If I enter a path in API domain or website domain it just gets ignored. I thought we were going to show an error to the user saying that they should instead use API base path / website base path right? Makes me wonder what else is broken...
  - It was using the SDK website / API domain normalization which didn't return the path part, but I did not remove it so it behaves like the SDK or API normalization would behave. I've updated it to show an error
- [ ]  Some sections that don't have the form in production, have the form in this PR - like "Setup the login UI" and "View the login UI (ReactJS only)" setup in [here](https://deploy-preview-176--admiring-bhabha-7b1be9.netlify.app/docs/passwordless/quick-setup/frontend). Why?
  - We used the appInfoForm in the same page already, and then displaying `/auth` in some parts regardless of what the user saved in the form was confusing, so it is updated in those sections. In the `Setup the login UI` section there was a comment in the React snippets that said `This renders the login UI on the /auth route`, which made no sense, so it was updated to show the `websiteBasePath`. Same was the case with the `View the login UI (ReactJS only)` section, where it was said that you can view the login UI by visiting `/auth`. What are your views? @rishabhpoddar

rishabhpoddar commented:
> What are your views?

Ah okay. Makes sense.


---------------
- [ ] If I let the path be `/auth`, it shows an error saying "Please enter a valid path". This is unexpected.

st-jay commented:
> > What are your views?
> 
> Ah okay. Makes sense.
> 
> * [ ]  If I let the path be `/auth`, it shows an error saying "Please enter a valid path". This is unexpected.

Updated

rishabhpoddar commented:
# Feedback
- [ ] If the api domian and the website domain (including the port) is the same (check that the normalised versions are the same), then the API base path cannot be a prefix of or equal to the website base path, and vice versa. In this case, you want to show an appropriate error.

st-jay commented:
> # Feedback
> * [ ]  If the api domian and the website domain (including the port) is the same (check that the normalised versions are the same), then the API base path cannot be a prefix of or equal to the website base path, and vice versa. In this case, you want to show an appropriate error.
> * [ ]  If I make API base path or website base path empty, it assumes `/auth` by default.. why? It should make those values `"/"` in the config.
> * [ ]  On a high level, can you explain the code flow? Based on your previous commit, ("only validate localstorage on first validation"), makes me think your code structure is not good. Either draw a diagram here or pseudo code (high level only)

* [X]  If the api domian and the website domain (including the port) is the same (check that the normalised versions are the same), then the API base path cannot be a prefix of or equal to the website base path, and vice versa. In this case, you want to show an appropriate error.
  * Updated
* [ ]  If I make API base path or website base path empty, it assumes `/auth` by default.. why? It should make those values `"/"` in the config.
  * Our reasoning was that as the fields are optional, users will leave the fields empty to use the default values, as opposed to actually leaving it empty for using the ‘/‘ route. Also, if the user doesn’t provide any `apiBasePath` or `websiteBasePath` value in the config, it defaults to `/auth`.  If the user explicitly enters `/` we display the ‘/‘ route.
*  [ ]  On a high level, can you explain the code flow? Based on your previous commit, ("only validate localstorage on first validation"), makes me think your code structure is not good. Either draw a diagram here or pseudo code (high level only)
   * Here's a high level explanation of how the code works - https://gist.github.com/st-jay/f231339e34ae80113b6247d5eed88b46
   * Here're all the test cases it currently handles - https://gist.github.com/st-jay/1e1f02e918ee69894ac446fb124ae32d

@rishabhpoddar

st-jay commented:
> If I make API base path or website base path empty, it assumes /auth by default.. why? It should make those values "/" in the config.

It's updated - when the apiBasePath and websiteBasePath fields get an empty value, they show the "/" value by default in the fields as well as the snippets.

rishabhpoddar commented:
# Feedback
- [ ] If I put `api.com/a` in apiBase path, it resolves as `/a.com/a`. Instead, it should display an error to the user

st-jay commented:
Updated the path validation to use the Node SDK normalisation logic for paths - https://github.com/supertokens/supertokens-node/blob/master/lib/ts/normalisedURLPath.ts

> [ ] If I put api.com/a in apiBase path, it resolves as /a.com/a. Instead, it should display an error to the user
- as per the SDK path normalisation, if the user enters `a.com/a`, it will be resolved to `/a`, instead of throwing an error. That is the current behaviour. Should we throw an error instead? @rishabhpoddar

st-jay commented:
### Feedback
- When the domains are equal and the paths are either equal or one path is a prefix to the other
  - Show the error message for the equal case and prefix case seperately
  - Show the error message for prefix in the correct field, i.e. if the `apiBasePath` is a prefix of `websiteBasePath` we show the error in apiBasePath field that it cannot be a prefix of the `websiteBasePath` and vice versa.
- For path validation, we need to first call the domain normalisation on the path and verify that it returns an empty string. If it does not return an empty string, we show an error.

st-jay commented:
> ### Feedback
> * When the domains are equal and the paths are either equal or one path is a prefix to the other
>   
>   * Show the error message for the equal case and prefix case seperately
>   * Show the error message for prefix in the correct field, i.e. if the `apiBasePath` is a prefix of `websiteBasePath` we show the error in apiBasePath field that it cannot be a prefix of the `websiteBasePath` and vice versa.
> * For path validation, we need to first call the domain normalisation on the path and verify that it returns an empty string. If it does not return an empty string, we show an error.

Updated to solve the above 2 points.

Also, found an issue where if the path was something like `.path/to/auth`, the domain normalisation didn't consider the `.path` part a domain, so didn't throw an error, and the path normalisation considered it a domain, so it removed it from the path. This issue is also fixed.
